src/Symfony/Component/Encryption/Sodium/SodiumKey.php

```diff
@@ -0,0 +1,140 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Encryption\Sodium;
+
+use Symfony\Component\Encryption\Exception\InvalidKeyException;
+use Symfony\Component\Encryption\KeyInterface;
+
+/**
+ * @internal
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+final class SodiumKey implements KeyInterface
+{
+    private ?string $secret = null;
+    private ?string $privateKey = null;
+    private ?string $publicKey = null;
+
+    /**
+     * A keypair can only be created from a public and private key.
+     */
+    private ?string $keypair = null;
+
+    public static function create(string $secret, string $keypair): self
+    {
+        $key = self::fromSecret($secret);
+        $key->keypair = $keypair;
+        $key->publicKey = sodium_crypto_box_publickey($keypair);
+        $key->privateKey = sodium_crypto_box_secretkey($keypair);
+
+        return $key;
+    }
+
+    /**
+     * The secret key length should be 32 bytes, but other sizes are accepted.
+     */
+    public static function fromSecret(string $secret): self
+    {
+        $key = new self();
+        if (\SODIUM_CRYPTO_SECRETBOX_KEYBYTES === \strlen($secret)) {
+            $key->secret = $secret;
+        } else {
+            // Trim the key to a good size
+            $key->secret = substr(sha1($secret), 0, \SODIUM_CRYPTO_SECRETBOX_KEYBYTES);
```

Comment:
First, [`sha1()` is not an appropriate function to use here](https://shattered.io).

But even worse, this is hex-encoded so you're getting roughly *half* of the expected entropy out of your secret from this sha1 invocation.

```diff
- $key->secret = substr(sha1($secret), 0, \SODIUM_CRYPTO_SECRETBOX_KEYBYTES);
+ $key->secret = mb_substr(hash('sha256', $secret, true), 0, \SODIUM_CRYPTO_SECRETBOX_KEYBYTES, '8bit');
```